flax/linen/module.py

```diff
@@ -538,6 +538,80 @@ def nowrap(fun: _CallableT) -> _CallableT:
   return fun
 
 
+def compact_name_scope(fun: _CallableT) -> _CallableT:
+  """Creates compact submodules from a method.
+
+  This is a decorator that allows you to define compact submodules from a
+  method. It's intention is to make it easier to port code Haiku code to Flax
+  by providing the same functionality.
+
+  Example::
+
+    >>> import flax.linen as nn
+    >>> import jax
+    >>> import jax.numpy as jnp
+    >>> from flax.core import pretty_repr
+    ...
+    >>> class Foo(nn.Module):
```

Comment:
Now that this is working so well (meaning we can call the method directly), this example is actually a bit confusing. It raises the question whether it perhaps is necessary to always have a `compact` method.

A simpler example could be:

```
class Foo(nn.Module):
  @nn.compact_name_scope
  def up(self, x):
    return nn.Dense(4)(x)

  @nn.compact_name_scope
  def down(self, x):
    return nn.Dense(3)(x)
```

This also makes it clear that we don't need any `compact` methods, and you could change the `init` call by adding `method=Foo.up`.

Then I think it would also be good to explain that `compact` and `compact_name_scope` can be mixed within a method, but there can only be one `compact` method.

Reply:
I see your point. I'll update the example a bit to make this distinction clearer, since the share the word `compact` in their name users might make that association based on this example.

Reply:
Awesome! Already approved it so feel free to submit :)